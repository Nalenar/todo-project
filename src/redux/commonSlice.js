/**
 * Redux слайс который содержит в себе всю логику связанную cо стандартными функциями
 * @module commonSlice
 * @function createSlice - создает слайс
 *
 * @param {string} name - Название
 * @param {array} initialState - Начальное состояние
 * @param {object} reducers - Редюсеры
 *
 * @param {object} state - состояние на данный момент
 * @param {object} action - данные действия
 *
 * @function toggleStyles - изменяет значение выбранного стиля
 * @function toggleWindows - изменяет значение выбранного окна
 * @function editTodo - передает id задачи в состояние
 */

import { createSlice } from "@reduxjs/toolkit";

export const commonSlice = createSlice({
  name: "common",
  initialState: {
    styles: {
      blur: "",
      theme: "light",
    },
    windows: {
      menu: false,
      settings: false,
      modal: false,
      editList: false,
      editTodo: false,
    },
    edit: {
      id: undefined,
    },
    search: undefined,
  },
  reducers: {
    toggleStyles: (state, action) => {
      state.styles[action.payload.style] = action.payload.string;
    },
    toggleWindows: (state, action) => {
      // if (action.payload.window === "menu") {
      //   state.windows = {
      //     menu: action.payload.toggle,
      //     settings: false,
      //     modal: false,
      //     editList: false,
      //     editTodo: false,
      //   };
      // } else if (action.payload.window === "settings") {
      //   state.windows = {
      //     menu: false,
      //     settings: action.payload.toggle,
      //     modal: false,
      //     editList: false,
      //     editTodo: false,
      //   };
      // } else {
      //   state.windows[action.payload.window] = action.payload.toggle;
      // }
      state.windows[action.payload.window] = action.payload.toggle;
    },
    editTodo: (state, action) => {
      state.edit.id = action.payload.id;
    },
    searchTodo: (state, action) => {
      state.search = action.payload.value;
    },
  },
});

/** @exports commonSlice.actions  - возможные действия со стандартными функциями */
export const { toggleStyles, toggleWindows, editTodo, searchTodo } =
  commonSlice.actions;

/**
 * @default
 * @exports commonSlice.reducer - экспортирует по-умолчанию редюсер всего слайса
 */
export default commonSlice.reducer;
