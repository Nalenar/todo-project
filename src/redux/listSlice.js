/**
 * Redux слайс который содержит в себе всю логику связанную cо списками задач
 * @module listSlice
 * @function createSlice - создает слайс
 *
 * @param {string} name - Название
 * @param {array} initialState - Начальное состояние
 * @param {object} reducers - Редюсеры
 *
 * @param {object} state - состояние на данный момент
 * @param {object} action - данные действия
 *
 * @function addCategory - создает список
 * @function deleteCategory - удаляет список
 * @function renameCategory - изменяет название списка на новое, указанное пользователем
 */

import { createSlice } from "@reduxjs/toolkit";

export const listSlice = createSlice({
  name: "lists",
  initialState: [
    {
      id: "HELP ME",
      title: "HELP ME",
    },
  ],
  reducers: {
    addCategory: (state, action) => {
      const list = {
        id: action.payload.title,
        title: action.payload.title,
      };
      state.push(list);
    },
    deleteCategory: (state, action) => {
      return state.filter((list) => list.id !== action.payload.id);
    },
    renameCategory: (state, action) => {
      const index = state.findIndex((list) => list.id === action.payload.id);
      state[index].title = action.payload.title;
    },
  },
});

/** @exports listSlice.actions  - возможные действия со списками */
export const { addCategory, deleteCategory, renameCategory } =
  listSlice.actions;

/**
 * @default
 * @exports listSlice.reducer - экспортирует по-умолчанию редюсер всего слайса
 */
export default listSlice.reducer;
