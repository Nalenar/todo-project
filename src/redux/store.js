/**
 * Главный файл который использует store
 * из библиотеки @reduxjs/toolkit
 * @module
 * 
 * Экспортирует объект хранящий в себе редюсеры слайсов
 * @exports 
 * 
 * @param {object} reducer - Содержит в себе название состояния и относящийся к нему редюсер
 * 
 * @param {object} todos - Содержит в себе все экшены, редюсеры и состояния для задач
 * @param {object} lists - Содержит в себе все экшены, редюсеры и состояния для списков 
 * @param {object} common - Содержит в себе все экшены, редюсеры и состояния для стандартных функций (например стилей, окон и т.д.)
 */

import { configureStore } from "@reduxjs/toolkit";
import todoReducer from "./todoSlice";
import commonReducer from "./commonSlice";
import listReducer from "./listSlice";

export default configureStore({
  reducer: {
    todos: todoReducer,
    lists: listReducer,
    common: commonReducer,
  },
});
