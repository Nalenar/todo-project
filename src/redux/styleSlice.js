import { createSlice } from "@reduxjs/toolkit";

export const styleSlice = createSlice({
  name: "blur",
  initialState: {
    blur: "",
  },
  reducers: {
    addBlur: (state) => {
      state.blur = "blur-sm";
    },
    deleteBlur: (state) => {
      state.blur = "";
    },
  },
});

export const { addBlur, deleteBlur } = styleSlice.actions;
export default styleSlice.reducer;
