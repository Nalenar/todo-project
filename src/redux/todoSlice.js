/**
 * Redux слайс который содержит в себе всю логику связанную с задачами
 * @module todoSlice
 * @function createSlice - Создает слайс
 *
 * @param {string} name - Название
 * @param {array} initialState - Начальное состояние
 * @param {object} reducers - Редюсеры
 *
 *
 * @param {object} state - состояние на данный момент
 * @param {object} action - данные действия
 *
 * @function addTodo - добавляет задание(объект) в состояние
 * @function toggleComplete - находит нужное задание меняет заначение completed
 * @function deleteTodo - удаляет нужное задание
 * @function toggleList - удаляет или добавляет список в состоянии
 */

import { createSlice } from "@reduxjs/toolkit";

export const todoSlice = createSlice({
  name: "todos",
  initialState: [
    {
      id: 1,
      title: "today",
      completed: true,
      list: ["today"],
    },
    {
      id: 2,
      title: "important",
      completed: true,
      list: ["today", "important"],
    },
  ],
  reducers: {
    addTodo: (state, action) => {
      const todo = {
        id: state.length + 1,
        title: action.payload.title,
        completed: false,
        list: [action.payload.list],
      };
      state.push(todo);
    },
    toggleComplete: (state, action) => {
      const index = state.findIndex((todo) => todo.id === action.payload.id);
      state[index].completed = action.payload.completed;
    },
    deleteTodo: (state, action) => {
      return state.filter((todo) => todo.id !== action.payload.id);
    },
    toggleList: (state, action) => {
      const index = state.findIndex((todo) => todo.id === action.payload.id);
      if (action.payload.toggle) {
        state[index].list.push(`${action.payload.list}`);
      } else {
        state[index].list = state[index].list.filter(
          (list) => list !== action.payload.list
        );
      }
    },
  },
});

/** @exports todoSlice.actions - экспортирует действия */

export const { addTodo, toggleComplete, deleteTodo, toggleList } =
  todoSlice.actions;

/**
 * @default
 *  @exports todoSlice.reducer - экспортирует по-умолчанию редюсер всего слайса
 */
export default todoSlice.reducer;
