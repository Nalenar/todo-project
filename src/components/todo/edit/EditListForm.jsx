import { useState } from "react";
import { useDispatch } from "react-redux";

import { renameCategory } from "../../../redux/listSlice";
import { toggleWindows } from "../../../redux/commonSlice";

const EditListForm = ({ id, title }) => {
  const [currentValue, setCurrentValue] = useState("");
  const dispatch = useDispatch();

  const handleEditFormSubmit = (event) => {
    event.preventDefault();
    dispatch(renameCategory({ id, title: currentValue }));
    setCurrentValue("");
    dispatch(toggleWindows({ window: "editList", toggle: false }));
  };

  return (
    <form
      onSubmit={handleEditFormSubmit}
      className="flex items-center justify-center"
    >
      <input
        autoFocus={true}
        type="text"
        placeholder={title}
        value={currentValue}
        onChange={(event) => setCurrentValue(event.target.value)}
        className="form-input mr-3 h-10 w-52 border-none bg-transparent p-1 text-center font-roboto text-2xl placeholder:text-blue-500 focus:ring-transparent focus:placeholder:text-gray-500"
      />
      <button
        type="submit"
        className="h-10 rounded-md bg-blue-500 px-2 py-1 font-roboto text-xl text-white shadow-md hover:bg-blue-600"
        role="button-edit_list"
      >
        Изменить
      </button>
    </form>
  );
};

export default EditListForm;
