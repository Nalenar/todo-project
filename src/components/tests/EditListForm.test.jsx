import { it, expect, describe, vi } from "vitest";
import { render, screen, fireEvent } from "@testing-library/react";
import * as reduxHooks from "react-redux";

import * as listActions from "../../redux/listSlice";
import * as commonActions from "../../redux/commonSlice";

import EditListForm from "../todo/edit/EditListForm";

vi.mock("react-redux");

const list = {
  id: "HELP ME",
  title: "HELP ME",
};

const mockedDispatch = vi.spyOn(reduxHooks, "useDispatch");

describe("EditListForm", () => {
  it("should create new list", () => {
    const component = render(<EditListForm {...list} />);

    expect(component).toMatchSnapshot();
  });

  it("should not rename list", () => {
    const dispatch = vi.fn();
    mockedDispatch.mockReturnValue(dispatch);

    const mockedRenameCategory = vi.spyOn(listActions, "renameCategory");
    const mockedToggleWindows = vi.spyOn(commonActions, "toggleWindows");

    render(<EditListForm {...list} />);

    fireEvent.submit(screen.getByRole("button-edit_list"));

    expect(dispatch).toHaveBeenCalledTimes(2);
    expect(mockedRenameCategory).toHaveBeenCalledWith({
      id: list.id,
      title: "",
    });
    expect(mockedToggleWindows).toHaveBeenCalledWith({
      window: "editList",
      toggle: false,
    });
  });
});
