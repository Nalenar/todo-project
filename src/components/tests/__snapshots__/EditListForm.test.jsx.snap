// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EditListForm > should create new list 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="flex items-center justify-center"
      >
        <input
          class="form-input mr-3 h-10 w-52 border-none bg-transparent p-1 text-center font-roboto text-2xl placeholder:text-blue-500 focus:ring-transparent focus:placeholder:text-gray-500"
          placeholder="HELP ME"
          type="text"
          value=""
        />
        <button
          class="h-10 rounded-md bg-blue-500 px-2 py-1 font-roboto text-xl text-white shadow-md hover:bg-blue-600"
          role="button-edit_list"
          type="submit"
        >
          Изменить
        </button>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="flex items-center justify-center"
    >
      <input
        class="form-input mr-3 h-10 w-52 border-none bg-transparent p-1 text-center font-roboto text-2xl placeholder:text-blue-500 focus:ring-transparent focus:placeholder:text-gray-500"
        placeholder="HELP ME"
        type="text"
        value=""
      />
      <button
        class="h-10 rounded-md bg-blue-500 px-2 py-1 font-roboto text-xl text-white shadow-md hover:bg-blue-600"
        role="button-edit_list"
        type="submit"
      >
        Изменить
      </button>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
